(GH-2321) Move plan_hierarchy docs to writing plans doc (#2347)
puppetlabs/bolt: documentation/writing_plans.md

types, such as `Optional[Sensitive]` or `Hash[String, Sensitive]`, will not be
automatically masked.

## Using Hiera data in plans

You can use the Puppet `lookup()` function in plans to look up Hiera data. It's useful to think
of looking up Hiera data in Bolt plans in two different contexts: inside [apply

> blocks](applying_manifest_blocks.md) and outside apply blocks. Inside apply blocks Bolt

Reviewer: ```suggestion
blocks](applying_manifest_blocks.md) and outside apply blocks. Inside apply blocks, Bolt
```

Author: Oh my gosh Tom literally had the same comment

compiles Puppet catalogs in a per-target context, and has unfettered access to Hiera data. You can
use the same Hiera config and data you would for Puppet, and look up data as you would expect
inside apply blocks. Outside apply blocks, Bolt is essentially executing a script. It doesn't have a
concept of a target or context, and thus cannot load per-target data. This breaks common Hiera
features like interpolations.

To look up static data outside of apply blocks, you can add a `plan_hierarchy` key to your Hiera
configuration. The `plan_hierarchy` key is specified at the same level as the `hierarchy` key and is
used whenever you look up data outside an apply block. By specifying both keys in the same Hiera
configuration, you can look up data inside and outside apply blocks in the same plan. This allows
you to use your existing Hiera configuration in Bolt plans without encountering an error if
interpolations exist and your plan tries to look up data outside an apply block.

> Using static Hiera data is particularly useful for user-specific data that you want to be looked up
> using Hiera (for example, to take advantage of Hiera eyaml). For example, with this Hiera

Reviewer: ```suggestion
Using static Hiera data is particularly useful for user-specific data that you want the plan to look up
using Hiera (for example, to take advantage of Hiera eyaml). For example, with this Hiera
```
Not sure if the plan is the thing doing the look up - mainly just trying to eliminate the passive

Author: I think saying the plan is doing the looking up makes sense.

configuration at `<PROJECT DIRECTORY>/hiera.yaml` plan users can set a user-specific API
key in the `plan_hierarchy.yaml` data file as well as use Hiera to look up a per-target filepath
inside an apply block.

> Using this hiera.yaml:

Reviewer: ```suggestion
Using this `hiera.yaml`:
```

```yaml
version: 5

hierarchy:
  - name: "Target specific data"
    path: "targets/%{trusted.certname}.yaml"
  - name: "Per-OS defaults"
    path: "os/%{facts.os.family}.yaml"
  - name: Common
    path: hierarchy.yaml

plan_hierarchy:
  - name: Common
    path: plan_hierarchy.yaml
```

With this data at `<PROJECT DIRECTORY>/data/plan_hierarchy.yaml`:
```
api_key: 12345
```

And this data at `<PROJECT DIRECTORY/data/targets/myhost.com`:
```
confpath: "C:\Program Files\Common Files\mytool.conf"
```

> Bolt will look up the API key in the first `lookup()` call, and the target-specific data inside the
> apply block:

Reviewer: ```suggestion
Bolt looks up the API key in the first `lookup()` call, and the target-specific data inside the
apply block:
```

types, such as `Optional[Sensitive]` or `Hash[String, Sensitive]`, will not be
automatically masked.

## Using Hiera data in plans

> You can also use the Puppet `lookup()` function in plans to lookup Hiera data. It's useful to think

Reviewer 2: ```suggestion
You can use the Puppet `lookup()` function in plans to lookup Hiera data. It's useful to think
```

of looking up Hiera data in Bolt plans in two different contexts: inside [apply

> blocks](applying_manifest_blocks.md), and outside apply blocks. Inside apply blocks Bolt is
> compiling Puppet catalogs in a per-target context, and has unfettered access to Hiera data. You can

Reviewer 2: ```suggestion
blocks](applying_manifest_blocks.md) and outside apply blocks. Inside apply blocks, Bolt
compiles Puppet catalogs in a per-target context and has unfettered access to Hiera data. You can
```

> use the same Hiera config and data you would for Puppet, and `lookup()` data as you would expect

Reviewer 2: ```suggestion
use the same Hiera config and data you would for Puppet, and look up data as you would expect
```

Author: Oo good call - changed this a few other places as well.

inside apply blocks. Outside apply blocks, Bolt is essentially executing a script. It doesn't have a
concept of a target or context, and thus cannot load per-target data. This breaks common Hiera
features like interpolations.

> In order to allow users to use their same Hiera configurations with Bolt to look up per-target data
> inside apply blocks in addition to being able to look up static data outside of apply blocks we
> created the `plan_hierarchy`. `plan_hierarchy` is a key you specify in your Hiera configuration file
> at the same level as `hierarchy` which loads static data that can be looked up by Bolt outside of
> apply blocks. If present, Bolt uses the data under the `plan_hierarchy` key outside of apply
> blocks, and uses the `hierarchy` key inside of apply blocks. This allows you to use your existing
> Hiera configs in Bolt plans without encountering an error if interpolation exists and your plan
> tries to look up data outside of an apply block.

Reviewer 2: ```suggestion
To look up static data outside of apply blocks, you can add a `plan_hierarchy` key to your Hiera
configuration. The `plan_hierarchy` key is specified at the same level as the `hierarchy` key
and is used whenever you look up data outside an apply block. By specifying both keys in the
same Hiera configuration, you can look up data inside and outside apply blocks in the same
plan. This allows you to use your existing Hiera configuration in Bolt plans without encountering
an error if interpolations exist and your plan tries to look up data outside an apply block.
```

Mostly trying to avoid "we created the `plan_hierarchy`" here.

> Using static Hiera data is particularly useful for user-specific data that you want to be looked up
> using Hiera (for example, to take advantage of Hiera eyaml). For example, with this Hiera
> configuration at `<PROJECT DIRECTORY>/hiera.yaml` plan users can set a user-specific API
> key in the `plan_hierarchy.yaml` data file as well as use Hiera to lookup a per-target filepath
> inside an apply block.

Reviewer 2: ```suggestion
Using static Hiera data is particularly useful for user-specific data that you want to look up
during a plan run. For example, with this Hiera configuration at
`<PROJECT DIRECTORY>/hiera.yaml`, you can set a user-specific API key in the
`plan_hierarchy.yaml` data file and use Hiera to lookup a per-target filepath
inside an apply block.
```
